src/rdkafka_queue.c

```diff
@@ -118,6 +124,33 @@ rd_kafka_q_t *rd_kafka_q_new0(rd_kafka_t *rk, const char *func, int line) {
         return rkq;
 }
 
+/*
+ * Handles the logic for increment/decrement of rkq_consumer_cnt.
+ * Only used internally when forwarding queues.
+ * @locks rd_kafka_q_lock(rkq)
+ */
+static void rd_kafka_q_consumer_cnt_propagate(rd_kafka_q_t *rkq,
+                                              int cnt_delta) {
+        rd_assert(rkq);
+        if (cnt_delta == 0)
+                return;
+
+        mtx_lock(&rkq->rkq_lock);
+        rkq->rkq_consumer_cnt += cnt_delta;
+        rd_assert(rkq->rkq_consumer_cnt >= 0);
+
+        if (!rkq->rkq_fwdq) {
+                mtx_unlock(&rkq->rkq_lock);
+                return;
+        }
+
+        /* Recursively propagate the change in the count. There is a chance of a
+         * deadlock here but only if the user is forwarding queues in a circular
+         * manner, which already is an incorrect mode of operation.*/
+        rd_kafka_q_consumer_cnt_propagate(rkq->rkq_fwdq, cnt_delta);
+        mtx_unlock(&rkq->rkq_lock);
```

Comment:
Shouldn't we release this lock earlier than above call? I think circular dependency case will cause a deadlock which will be very hard to identify even if it is a very rare scenario.

Reply:
That is not possible to do, I think
Consider the case that A, B are queues.. Someone calls forward(A,B). propagate(A) is called. After processing, suppose we unlock the mutex and then call propagate(B). 
Before propagate(B) starts to run, someone does forward(A, NULL) on another thread. If A->rkq_fwdq is the only reference to B, then B will get destroyed now, and we might encounter a segmentation fault on trying to access fields of B when propagate(B) finally runs.

Both these cases (circular dependency and concurrent calling of forward(A,B) and forward(A,NULL)) are rare, but while causing a circular dependency is an error case which the user should take care to avoid, concurrent forwarding is legitimate use case if someone is trying something with dynamic queue forwarding. So better to support the legitimate use case since we have to choose between them.

Changed the documentation for this.